File: WorkcraftCore/src/main/java/org/workcraft/dom/visual/ColorisableGraphicalContent.java
package org.workcraft.dom.visual;

import java.awt.Graphics2D;

import org.workcraft.gui.graph.tools.Colorisation;
import org.workcraft.util.Function;
import org.workcraft.util.Function2;

public interface ColorisableGraphicalContent {
	ColorisableGraphicalContent EMPTY = new ColorisableGraphicalContent(){
		@Override
		public void draw(DrawRequest request) {
		}
	};
	
	public class Util {
		public static ColorisableGraphicalContent fromFunc(final Function<Colorisation, GraphicalContent> func) {
			return new ColorisableGraphicalContent() {
				@Override
				public void draw(DrawRequest request) {
					func.apply(request.getColorisation()).draw(request.getGraphics());
				}
			};
		}
		
		public static GraphicalContent applyColourisation(final ColorisableGraphicalContent content, final Colorisation colour) {
			return new GraphicalContent() {
				
				@Override
				public void draw(final Graphics2D graphics) {
					content.draw(new DrawRequest() {
						@Override
						public Graphics2D getGraphics() {
							return graphics;
						}
						
						@Override
						public Colorisation getColorisation() {
							return colour;
						}
					});
				}
			};
		}
		public final static Function2<ColorisableGraphicalContent, Colorisation, GraphicalContent> applyColourisationFunc = new Function2<ColorisableGraphicalContent, Colorisation, GraphicalContent>() {
			@Override
			public GraphicalContent apply(ColorisableGraphicalContent argument1, Colorisation argument2) {
				return applyColourisation(argument1, argument2);
			}
		};
	}

	public void draw(DrawRequest request);
}

File: Graphics/src/main/scala/org/workcraft/graphics/ColorisableGraphicalContent.scala
package org.workcraft.graphics
import java.awt.Graphics2D
import java.awt.geom.AffineTransform

trait ColorisableGraphicalContent {
  def draw(request: DrawRequest)
  
  def applyColorisation(color: Colorisation) = new GraphicalContent {
    def draw (graphics: Graphics2D) = ColorisableGraphicalContent.this.draw (DrawRequest(graphics, color))
  }

  def transform (transformation: AffineTransform) = 
    ColorisableGraphicalContent (applyColorisation(_).transform(transformation))
  
  def compose (top: ColorisableGraphicalContent) =
    ColorisableGraphicalContent (c => applyColorisation(c).compose(top.applyColorisation(c)))
}

object ColorisableGraphicalContent {
	val Empty = new ColorisableGraphicalContent {
		def draw(request: DrawRequest) = {}
	}
	
	def apply (f: Colorisation => GraphicalContent) = new ColorisableGraphicalContent {
	  def draw (request: DrawRequest) = f(request.colorisation).draw(request.graphics)
	}
}